main.c

```diff
@@ -122,28 +144,63 @@ int main(int argc, char *argv[]) {
     	size_t len;
     	struct stat s;
 	uint64_t total = 0;
-	struct timespec start_time;
 
   	cmd_line_options arguments = (cmd_line_options){.file_name = "", .q = 8, .c = 1};
-  	argp_parse (&argp, argc, argv, 0, 0, &arguments);
+	int ret = argp_parse (&argp, argc, argv, 0, 0, &arguments);
+
+	if (ret != 0) {
+		fprintf (stderr, "Unexpected error while parsing command line arguments: %s\n",
+			 strerror (ret));
+		return 1;
+	}
 	
 	int nthreads = arguments.c;
 	threads = malloc(nthreads * sizeof(pthread_t));
 	tasks = malloc(nthreads * sizeof(lua_task));
 
-	clock_gettime(CLOCK_MONOTONIC, &start_time);
-	double start_time_ns = start_time.tv_sec * 1e9 +  start_time.tv_nsec;
+	if (threads == NULL || tasks == NULL) {
+		fprintf(stderr, "Unable to allocate memory: %s\n", strerror (errno));
+		return 2;
+	}
 
 	for (i = 0; i < nthreads; i++) {
-		tasks[i] = (lua_task){.file = arguments.file_name, .start_ns = start_time_ns};
-		pthread_create(&threads[i], NULL, wrapper, &tasks[i]);
+		tasks[i] = (lua_task){.file = arguments.file_name};
+		int ret = pthread_create(&threads[i], NULL, wrapper, &tasks[i]);
+
+		if (ret != 0) {
+			fprintf (stderr, "Cannot create thread(%d): %s\n", i,
+				 strerror (errno));
+			return 3;
+		}
 	}
 
+	uint64_t start_time_ns = 0;
+	uint64_t end_time_ns = 0;
+
   	for (i = 0; i < nthreads; i++) {
-    		pthread_join(threads[i], NULL);
+		int ret = pthread_join(threads[i], NULL);
+
+		if (ret != 0) {
+			fprintf (stderr, "Thread join failed(%d): %s\n", i,
+				 strerror (errno));
+			return 4;
+		}
+
 		total += tasks[i].ctr;
+		/* Select the earliest starting time and the latest end time. */
+		if (start_time_ns == 0) {
+			start_time_ns = tasks[i].start_ns;
+			end_time_ns = tasks[i].end_ns;
+		}
+		else {
+			start_time_ns = MIN (start_time_ns, tasks[i].start_ns);
```

Comment:
I am not entirely convinced that this method is much more accurate. Maybe the more accurate way would be to compute the average between all threads?

Reply:
Average will give you a throughput rate per thread, which is not incorrect, but definitely a different thing from total throughput.  If you want total throughput you either get start and end time in the main thread or you get them in individual threads and get the earliest start time and latest end time.  The latter eliminates some amount of the cloning and joining latency, although it shouldn't make a significant difference if the benchmark environment is controlled.